pkg/container/versioned/value.go

```diff
@@ -0,0 +1,510 @@
+	newNode.versions.past.store(invalidVersion)
+
+	// find if there is a current value that is valid for this new version
+	node := v.head.Load()
+	prev := &v.head
+	for node != nil {
+		if version < node.versions.first {
+			return fmt.Errorf("SetAt may not add values at versions lower than those already added (%d<%d): %w", version, node.versions.first, ErrStaleVersion)
+		}
+
+		if node.versions.contains(version) {
+			// link the new node after the current one
+			newNode.next.Store(node.next.Load())
+			node.next.Store(newNode)
+
+			// truncate the validity of this node to end at 'version' *after* the new
+			// node with validity starting from 'version' has been linked after it
+			// (above), so that either this or the new value is reachable at all times
+			// for readers with 'version'
+			node.versions.past.store(version)
+			break
+		}
+
+		node = node.next.Load()
+		if node != nil {
+			prev = &node.next
+		}
+	}
+	if node == nil {
+		// Add the new value at the end
+		prev.Store(newNode)
+	}
+
+	return nil
+}
+
+// RemoveAt changes the validity of the stored value previously valid at version 'next' to have
+// ended at version 'next'.
+// 'next' must be later than any the current version visible to the readers.
+// Returns an error if this is not the case.
+// Callers must coordinate for mutual exclusion.
+func (v *Value[T]) RemoveAt(tx *Tx) error {
+	version := tx.nextVersion
+	if version == invalidVersion {
+		return ErrInvalidVersion
+	}
+
+	for node := v.head.Load(); node != nil; node = node.next.Load() {
+		if version < node.versions.first {
+			return fmt.Errorf("RemoveAt may not be called with version lower than existing already (%d<%d): %w", version, node.versions.first, ErrStaleVersion)
+		}
+
+		if node.versions.contains(version) {
+			// Truncate the validity of this node to end at 'version'.
+			// After this readers with 'version' and above no longer see this value,
+			// while readers with versions before 'version' still see this.
+			node.versions.past.store(version)
+			break
+		}
+	}
+	return nil
+}
+
+// RemoveBefore removes all values whose validity ends before 'keepVersion'.
+// Caller must coordinate for mutual exclusion.
+func (v *Value[T]) RemoveBefore(keepVersion KeepVersion) {
+	version := version(keepVersion)
+	// find all values that are no longer valid at 'version'
+	node := v.head.Load()
+	for node != nil && node.versions.past.load() <= version {
+		// This node is no longer visible for readers with 'version' and above,
+		// so this can be safely removed.
+		node = node.next.Load()
+	}
+	v.head.Store(node)
+}
+
+// At returns value of type 'T' valid for the given version, or an empty value if none is found.
+func (v *Value[T]) At(handle *VersionHandle) (empty T) {
+	if handle != nil {
+		version := handle.version
+		for node := v.head.Load(); node != nil; node = node.next.Load() {
+			if node.versions.contains(version) {
+				return node.value
+			}
+		}
+	}
+	return empty
+}
```

Comment:
The return value named `empty` will look confusing in godoc because it's not always empty. Let's get rid of the name:

```suggestion
func (v *Value[T]) At(handle *VersionHandle) T {
	if handle != nil {
		version := handle.version
		for node := v.head.Load(); node != nil; node = node.next.Load() {
			if node.versions.contains(version) {
				return node.value
			}
		}
	}
	var empty T
	return empty
}
```

Reply:
Done, thanks!